crates/bevy_ui/src/geometry.rs

```diff
@@ -255,15 +255,19 @@ pub enum ValArithmeticError {
 }
 
 impl Val {
-    /// Resolves a [`Val`] from the given context values and returns this as an [`f32`].
-    /// The [`Val::Px`] value (if present), `parent_size` and `viewport_size` should all be in the same coordinate space.
-    /// Returns a [`ValArithmeticError::NonEvaluable`] if the [`Val`] is impossible to resolve into a concrete value.
+    /// Resolves this [`Val`] to a value in physical pixels from the given `scale_factor`, `parent_size`,
+    /// and `viewport_size`.
     ///
-    /// **Note:** If a [`Val::Px`] is resolved, its inner value is returned unchanged.
-    pub fn resolve(self, parent_size: f32, viewport_size: Vec2) -> Result<f32, ValArithmeticError> {
+    /// Returns a [`ValArithmeticError::NonEvaluable`] if the [`Val`] is impossible to resolve into a concrete value.
+    pub fn resolve(
+        self,
+        scale_factor: f32,
```

Comment:
I'm not familiar with UI code and it took me a second to work out what `scale_factor` did. Maybe worth a rename and perhaps a comment mentioning it's related to physical/logical.

```suggestion
        pixel_scale_factor: f32,
```

Reply:
Yep scale factor has always been a bit difficult to deal with. 
There are multiple scale factors: the scale factor(s) of the render targets and also Bevy's global UI scale factor which is set using the `UiScale` resource. So if you create a UI node with `width: Val::Px(x)` then (assuming there's no scaling by the transform and the layout doesn't place any other constraints on the width) the final width in physical pixels is calculated as `x * render_target_scale_factor * ui_scale`.